### PR TITLE
Optimize async trampoline, add ContextSwitch

### PR DESCRIPTION
**Pull request by alexandru**
Adding some optimizations that I've also pushed in Monix's master. 
Work in progress.

- add `ContextSwitch` in `IO`'s ADT definition for changing the `IOConnection`
- optimize `uncancelable` based on it
- we might express `bracket` in terms of `ContextSwitch` as well, but I'm not sure if it's needed
- replace Scala's `ArrayStack` with a custom one, because Scala's grows indefinitely, without releasing memory back
- optimize `TrampolineEC` with an implementation that minimizes reads from `ThreadLocal` and that uses the provided `ArrayStack`
- optimize `race`, `racePair`, `start`, `parMap2` to avoid extraneous forks

Also fixed in this PR:

- `race`, `racePair` and `parMap2` are building a stack unsafe cancelable, which can be witnessed when building a task from a large list, two by two; the solution being to trampoline those as well

Some numbers:

| Benchmark | 1.0.0-RC-8ed6e71 | This PR |
|--------------|------:|-------------:|
| AsyncBenchmark.async  | 166143.881 | 158502.277 |
| AsyncBenchmark.bracket | 68624.565 | 91128.723 | 
| AsyncBenchmark.cancelBoundary | 91922.709 | 111832.226 |
| AsyncBenchmark.cancelable  | 134808.349 | 134175.404 |
| AsyncBenchmark.parMap2 | 5821.615 | 13761.070 |
| AsyncBenchmark.race  | 22520.568 | 36113.413 | 
| AsyncBenchmark.racePair  |  20030.172  | 32882.632 | 
| AsyncBenchmark.start   |  2238.351  | 3470.732  | 
| AsyncBenchmark.uncancelable |   115187.834  |  492934.556 | 

Explanation:

- `IO.async` is the same
- `IO.bracket` benefits from the improvements in `onCancelRaiseError` — this will be good once we make `acquire` and `release` to be non-cancelable (PR #241)
- `IO.cancelBoundary` is better due to improvements in `RestartCallback`
- `IO.cancelable` is the same
- `IO.parMap2`, `race`, `racePair`, `start` are better due to eliminating extraneous forks, plus the improvements in `RestartCallback`
- `IO#uncancelable` is better due to `ContextSwitch` and this is very important for the performance of `bracket`, otherwise we'll take a hit when #241 is merged

### CODE REVIEW

**Comment by alexandru** [comment]
Also added some numbers to go with these improvements.

File: core/shared/src/main/scala/cats/effect/internals/ArrayStack.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    }
+    array(index) = a.asInstanceOf[AnyRef]
+  }
+
+  /** Pushes an entire iterator on the stack. */
+  def pushAll(cursor: Iterator[A]): Unit = {
+    while (cursor.hasNext) push(cursor.next())
+  }
+
+  /** Pushes an entire sequence on the stack. */
```
Comment (rossabaker) [inline core/shared/src/main/scala/cats/effect/internals/ArrayStack.scala]: Nit: "entire iterable"

File: core/shared/src/main/scala/cats/effect/internals/ArrayStack.scala [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  def pushAll(seq: Iterable[A]): Unit = {
+    pushAll(seq.iterator)
+  }
+
+  /** Pushes an entire sequence on the stack. */
```
Comment (rossabaker) [inline core/shared/src/main/scala/cats/effect/internals/ArrayStack.scala]: Nit: "entire array stack"

File: core/shared/src/main/scala/cats/effect/internals/IOConnection.scala [hunk 3]
```diff
@@ -54,6 +54,16 @@ private[effect] sealed abstract class IOConnection {
    */
   def push(cancelable: Cancelable): Unit
 
+  /**
+   * Pushes a pair of `IOConnection` on the stack, which on
+   * cancelation will get trampolined.
```
Comment (rossabaker) [inline core/shared/src/main/scala/cats/effect/internals/IOConnection.scala]: cancellation... there are a bunch of these again. :us: :trollface:

File: core/shared/src/main/scala/cats/effect/internals/IOConnection.scala [hunk 4]
```diff
@@ -139,4 +156,17 @@ private[effect] object IOConnection {
     def tryReactivate(): Boolean =
       state.compareAndSet(null, Nil)
   }
+
+  private final class TrampolinedPair(lh: IOConnection, rh: IOConnection)
+    extends Cancelable.Type with Runnable {
+
+    override def run(): Unit =
+      try lh.cancel()
+      finally rh.cancel()
```
Comment (rossabaker) [inline core/shared/src/main/scala/cats/effect/internals/IOConnection.scala]: We'll swallow an error here if they both throw. Is that okay?

File: core/shared/src/main/scala/cats/effect/internals/IOForkedStart.scala [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+   */
+  def apply[A](task: IO[A], timer: Timer[IO]): IO[A] =
+    if (detect(task)) task
+    else timer.shift.flatMap(_ => task)
+
+  /**
+   * Returns `true` if the given task is known to fork execution,
+   * or `false` otherwise.
+   */
+  @tailrec def detect(task: IO[_], limit: Int = 8): Boolean = {
```
Comment (rossabaker) [inline core/shared/src/main/scala/cats/effect/internals/IOForkedStart.scala]: Why 8?

File: core/shared/src/main/scala/cats/effect/internals/IOParMap.scala [hunk 6]
```diff
@@ -14,104 +14,115 @@
...
+    conn: IOConnection,
+    cb: Callback.T[C])
+    extends Runnable {
+
+    /**
+     * State synchronized by an atomic reference. Possible values:
+     *
+     *  - null: none of the 2 tasks have finished yet
+     *  - Left(a): the left task is waiting for the right one
+     *  - Right(a): the right task is waiting for the left one
```
Comment (rossabaker) [inline core/shared/src/main/scala/cats/effect/internals/IOParMap.scala]: Nit: `Right(b)`, no?

File: core/shared/src/main/scala/cats/effect/internals/IORunLoop.scala [hunk 7]
```diff
@@ -220,17 +229,14 @@ private[effect] object IORunLoop {
...
     bFirst: Bind,
-    bRest: CallStack,
-    register: (IOConnection, Either[Throwable, Any] => Unit) => Unit): IO[A] = {
+    bRest: CallStack): IO[A] = {
 
     // Hitting an async boundary means we have to stop, however
     // if we had previous `flatMap` operations then we need to resume
     // the loop with the collected stack
-    if (bFirst != null || (bRest != null && bRest.nonEmpty))
+    if (bFirst != null || (bRest != null && !bRest.isEmpty))
```
Comment (rossabaker) [inline core/shared/src/main/scala/cats/effect/internals/IORunLoop.scala]: Curiosity: is this known to be faster in practice?  (I don't have a problem with it.)

File: core/shared/src/main/scala/cats/effect/internals/IOForkedStart.scala (hunk 5 above)
Comment (alexandru) [reply core/shared/src/main/scala/cats/effect/internals/IOForkedStart.scala]: When allocating memory it's good if it's a power of two and 8 is the size of a CPU cache line on Intel's Core i7.

File: core/shared/src/main/scala/cats/effect/internals/IOParMap.scala (hunk 6 above)
Comment (alexandru) [reply core/shared/src/main/scala/cats/effect/internals/IOParMap.scala]: Right.

File: core/shared/src/main/scala/cats/effect/internals/IORunLoop.scala (hunk 7 above)
Comment (alexandru) [reply core/shared/src/main/scala/cats/effect/internals/IORunLoop.scala]: It's not faster, but I just replaced Scala's ArrayStack with my own, simpler implementation that doesn't have `nonEmpty` 🙂

**Comment by alexandru** [comment]
Merging this as we need time for testing it, the internals of the run-loop being kind of dirty.

**Pull request merged**